sql/core/src/test/scala/org/apache/spark/sql/UserDefinedTypeSuite.scala

```diff
@@ -71,7 +73,104 @@ object UDT {
 
 }
 
+// object and classes to test SPARK-19311
+
+// Trait/Interface for base type
+@SQLUserDefinedType(udt = classOf[ExampleBaseTypeUDT])
+sealed trait IExampleBaseType extends Serializable {
+  def field: Int
+}
+
+// Trait/Interface for derived type
+@SQLUserDefinedType(udt = classOf[ExampleSubTypeUDT])
+sealed trait IExampleSubType extends IExampleBaseType
+
+// a base class
+class ExampleBaseClass(override val field: Int) extends IExampleBaseType {
+  override def toString: String = field.toString
+
+}
+
+// a derived class
+class ExampleSubClass(override val field: Int)
+  extends ExampleBaseClass(field) with IExampleSubType
+
+// UDT for base class
+private[spark] class ExampleBaseTypeUDT extends UserDefinedType[IExampleBaseType] {
+
+  override def sqlType: StructType = {
+    StructType(Seq(
+      StructField("intfield", IntegerType, nullable = false)))
+  }
+
+  override def serialize(obj: IExampleBaseType): InternalRow = {
+    val row = new GenericInternalRow(1)
+    row.setInt(0, obj.field)
+    row
+  }
+
+  override def deserialize(datum: Any): IExampleBaseType = {
+    datum match {
+      case row: InternalRow =>
+        require(row.numFields == 1,
+          s"VectorUDT.deserialize given row with length " +
+            s"${row.numFields} but requires length == 1")
+        val field = row.getInt(0)
+        new ExampleBaseClass(field)
+    }
+  }
+
+  override def userClass: Class[IExampleBaseType] = classOf[IExampleBaseType]
+
+  override def hashCode(): Int = classOf[ExampleBaseTypeUDT].getName.hashCode()
+
+  override def equals(other: Any): Boolean = other.isInstanceOf[ExampleBaseTypeUDT]
+
+  override def typeName: String = "exampleBaseType"
+
+  private[spark] override def asNullable: ExampleBaseTypeUDT = this
```

Comment:
Nit: to simplify the test cases, please remove `asNullable `, `typeName `, `equals` and `hashCode`